pkg/proxy/runner/bounded_frequency_runner.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runner
+
+import (
+	"fmt"
+	"time"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+)
+
+// BoundedFrequencyRunner manages runs of a user-provided work function.
+type BoundedFrequencyRunner struct {
+	name string // the name of this instance
+
+	minInterval   time.Duration // the min time between runs
+	retryInterval time.Duration // the time between a run and a retry
+	maxInterval   time.Duration // the max time between runs
+
+	run chan struct{} // try an async run
+
+	fn               func() error // the work function
+	minIntervalTimer clock.Timer
+	nextRunTimer     clock.Timer // Combined timer for maxInterval and retryInterval logic
+	clock            clock.Clock
+}
+
+// NewBoundedFrequencyRunner creates and returns a new BoundedFrequencyRunner.
+// This runner manages the execution frequency of the provided function `fn`.
+//
+// The runner guarantees two properties:
+//  1. Minimum Interval (`minInterval`): At least `minInterval` must pass between
+//     the *completion* of one execution and the *start* of the next. Calls to
+//     `Run()` during this cooldown period are coalesced and deferred until the
+//     interval expires. This prevents burst executions.
+//  2. Maximum Interval (`maxInterval`): The function `fn` is guaranteed to run
+//     at least once per `maxInterval`, ensuring periodic execution even without
+//     explicit `Run()` calls (e.g., for refreshing state).
+//
+// `maxInterval` must be greater than or equal to `minInterval`; otherwise,
+// this function will panic.
+//
+// If `fn` returns an error, then it will be run again no later than `retryInterval`
+// (unless another trigger, like `Run()` or `maxInterval`, causes it to run sooner). Any
+// successful run will abort the retry attempt.
+func NewBoundedFrequencyRunner(name string, fn func() error, minInterval, retryInterval, maxInterval time.Duration) *BoundedFrequencyRunner {
```

Comment:
I notice that bursting went away - IIRC the main reason for that was that a new service creation gets two different events - the Service and the EndpointSlice - and without a burst we would ALWAYS be adding minInterval startup latency for every service.

Am I misremembering?